### PR TITLE
feat(request) support get_header from ffi

### PR DESCRIPTION
**Pull request by Suika-Kong**
This is a part of the solution for FT-2780.
The API will be later used in the PDK.

See also:
https://github.com/Kong/kong/pull/8708

### CODE REVIEW

File: lualib/resty/kong/request.lua [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+local ffi_str = ffi.string
+
+local subsystem = ngx.config.subsystem
+
+local get_header
+if subsystem == "http" then
+    ffi.cdef[[
+        ngx_str_t *ngx_http_lua_kong_request_get_header(ngx_http_request_t *r, ngx_str_t name, size_t search_limit);
+    ]]
+    function get_header(name, limit)
```
Comment (chronolaw) [inline lualib/resty/kong/request.lua:16]: `get_header = functioin()` may be  a better style.

File: lualib/resty/kong/request.lua [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        local r = get_request()
+        local name_ffi = ffi_new("ngx_str_t")
```
Comment (chronolaw) [inline lualib/resty/kong/request.lua:18]: `name_ffi` can be cached in module scope.

File: lualib/resty/kong/request.lua [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        name_ffi.data = name
+        name_ffi.len = #name
+        local ret = C.ngx_http_lua_kong_request_get_header(r, name_ffi, limit or 100)
```
Comment (chronolaw) [inline lualib/resty/kong/request.lua:21]: `100` should be defined as a constant.

File: lualib/resty/kong/request.lua [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        if ret ~= nil then
+            return ffi_str(ret.data, ret.len)
+        end
+    end
+end
+
+return {
+    get_header = get_header
```
Comment (chronolaw) [inline lualib/resty/kong/request.lua:29]: We should always have a `,` at each line's end.

File: src/ngx_http_lua_kong_request.c [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+#include <ctype.h>
+#include <stddef.h>
+#include <ngx_config.h>
+#include <ngx_core.h>
+#include <ngx_http.h>
+
+// name should be lowercased and preprocessed for both search functions
+
+// by hash
+static ngx_str_t *search_known_header(ngx_http_request_t *r, ngx_str_t name) {
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:11]: Please use nginx C style, `{` should in a new line.

File: src/ngx_http_lua_kong_request.c [hunk 6]
```diff
@@ -0,0 +1,101 @@
+
+#include <ctype.h>
+#include <stddef.h>
+#include <ngx_config.h>
+#include <ngx_core.h>
+#include <ngx_http.h>
+
+// name should be lowercased and preprocessed for both search functions
+
+// by hash
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:10]: We should use C style comments, that is `/* */`

File: src/ngx_http_lua_kong_request.c [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            i = 0;
+        }
+
+        if (header[i].hash == 0) {
+            continue;
+        }
+
+        size_t n;
+        for (n = 0u; n < name.len && n < header[i].key.len; n++) {
+            u_char ch = tolower(header[i].key.data[n]);
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:57]: Use `ngx_tolower`

File: src/ngx_http_lua_kong_request.c [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    return ret;
+}
+
+// search request header named "name"
+// and within search_limit
+ngx_str_t *ngx_http_lua_kong_request_get_header(ngx_http_request_t *r, ngx_str_t name, size_t search_limit) {
+    ngx_str_t processed_name = header_preprocess(r, name);
+    ngx_str_t * ret = search_known_header(r, processed_name);
+    if (!ret) {
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:97]: We should compare to `NULL` in Nginx style.

**Comment by Suika-Kong** [comment]
new solution: https://github.com/Kong/lua-kong-nginx-module/pull/39

File: lualib/resty/kong/request.lua [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        ngx_str_t *ngx_http_lua_kong_request_get_header(
+            ngx_http_request_t *r, ngx_str_t name, size_t search_limit);
+    ]]
+
+    local DEFAULT_HEADER_LIMIT = 100
+
+    local name_ffi = ffi_new("ngx_str_t")
+    get_header = function (name, limit)
+        local r = get_request()
+        name_ffi.data = name
```
Comment (chronolaw) [inline lualib/resty/kong/request.lua:23]: I think `name_ffi` should be changed to a simple `name_str`. `ffi` is a bit verbose.

File: lualib/resty/kong/request.lua [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        name_ffi.len = #name
+        local ret =
+            C.ngx_http_lua_kong_request_get_header(r, name_ffi, limit or DEFAULT_HEADER_LIMIT)
+        if ret ~= nil then
+            return ffi_str(ret.data, ret.len)
+        end
```
Comment (chronolaw) [inline lualib/resty/kong/request.lua:29]: How about `ret == nil`, should we return an error message?

File: lualib/resty/kong/request.lua (hunk 9 above)
Comment (Suika-Kong) [inline lualib/resty/kong/request.lua:23]:
```suggestion
        name_str.data = name
```

File: lualib/resty/kong/request.lua [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+local get_header
+if subsystem == "http" then
+    ffi.cdef[[
+        ngx_str_t *ngx_http_lua_kong_request_get_header(
+            ngx_http_request_t *r, ngx_str_t name, size_t search_limit);
+    ]]
+
+    local DEFAULT_HEADER_LIMIT = 100
+
+    local name_ffi = ffi_new("ngx_str_t")
```
Comment (Suika-Kong) [inline lualib/resty/kong/request.lua:20]:
```suggestion
    local name_str = ffi_new("ngx_str_t")
```

File: lualib/resty/kong/request.lua [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    get_header = function (name, limit)
+        local r = get_request()
+        name_ffi.data = name
+        name_ffi.len = #name
```
Comment (Suika-Kong) [inline lualib/resty/kong/request.lua:24]:
```suggestion
        name_str.len = #name
```

File: lualib/resty/kong/request.lua (hunk 10 above)
Comment (Suika-Kong) [reply lualib/resty/kong/request.lua:29]: when it's nil it simply means we don't have such a header(which has the same behaviour like `ngx.var.http_`.

File: lualib/resty/kong/request.lua [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        local ret =
+            C.ngx_http_lua_kong_request_get_header(r, name_ffi, limit or DEFAULT_HEADER_LIMIT)
```
Comment (Suika-Kong) [inline lualib/resty/kong/request.lua:26]:
```suggestion
            C.ngx_http_lua_kong_request_get_header(r, name_str, limit or DEFAULT_HEADER_LIMIT)
```

File: src/ngx_http_lua_kong_request.c [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+#include <ngx_http.h>
+
+/* name should be lowercased and preprocessed for both search functions */
+
+/* by hash */
+static ngx_str_t *search_known_header(ngx_http_request_t *r, ngx_str_t name)
+{
+    /* Calculate a hash of lowercased header name */
+    ngx_uint_t hash = 0;
+    for (size_t i = 0; i < name.len; i++) {
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:12]: How about the api `ngx_hash_key`, which seems to have the same effect.

File: src/ngx_http_lua_kong_request.c [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        hash = ngx_hash(hash, name.data[i]);
+    }
+
+    ngx_http_core_main_conf_t *cmcf =
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:16]: As nginx C style, we should put declaration in the front.

File: src/ngx_http_lua_kong_request.c [hunk 16]
```diff
@@ -0,0 +1,109 @@
+#include <ngx_config.h>
+#include <ngx_core.h>
+#include <ngx_http.h>
+
+/* name should be lowercased and preprocessed for both search functions */
+
+/* by hash */
+static ngx_str_t *search_known_header(ngx_http_request_t *r, ngx_str_t name)
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:8]: Also nginx style, functon's return type `static ngx_str_t*` should be at a single line.

File: src/ngx_http_lua_kong_request.c [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+        struct (hh->offset tells in which one). */
+    return &((*(ngx_table_elt_t **)
+        ((char *) &r->headers_in + hh->offset))->value);
+}
+
+/* linear search */
+static ngx_str_t *search_unknown_header(
+    ngx_http_request_t *r, ngx_str_t name, size_t search_limit)
+{
+    ngx_list_part_t *part = &(r->headers_in.headers.part);
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:41]: Still C style, not C++ Style, declaration then assign them.

File: src/ngx_http_lua_kong_request.c [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    ngx_table_elt_t *header = part->elts;
+
+    /* not limit when search_limit == 0 */
+    for (size_t i = 0u; search_limit == 0 || i < search_limit; i++) {
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:45]: `size_t i` is C++ style, change it to C style to fit nginx.

File: src/ngx_http_lua_kong_request.c [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+            part = part->next;
+            header = part->elts;
+            i = 0;
+        }
+
+        if (header[i].hash == 0) {
+            continue;
+        }
+
+        size_t n;
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:60]: C style

File: src/ngx_http_lua_kong_request.c [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+        }
+    }
+
+    return NULL;
+}
+
+static ngx_str_t header_preprocess(
+    ngx_http_request_t *r, ngx_str_t name)
+{
+    ngx_str_t ret;
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:84]: `ret` is not a good name, use a more readable name.

File: src/ngx_http_lua_kong_request.c [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+    ret.data = ngx_palloc(r->pool, name.len);
+    ret.len = name.len;
+
+    for (size_t n = 0u; n < name.len; ++n) {
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:88]: `n` maybe `i`?

Comment (chronolaw) [reply src/ngx_http_lua_kong_request.c:88]: How about `ngx_strlow`?

File: src/ngx_http_lua_kong_request.c [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    return ret;
+}
+
+/* search request header named "name" and within search_limit */
+ngx_str_t * ngx_http_lua_kong_request_get_header(
+        ngx_http_request_t *r, ngx_str_t name, size_t search_limit)
+{
+    ngx_str_t processed_name = header_preprocess(r, name);
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:103]: C style.

File: src/ngx_http_lua_kong_request.c [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+    ngx_str_t * ret = search_known_header(r, processed_name);
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:104]: `ret`, use a more friendly name.

File: lualib/resty/kong/request.lua [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+    ]]
+
+    local DEFAULT_HEADER_LIMIT = 100
+
+    local name_ffi = ffi_new("ngx_str_t")
+    get_header = function (name, limit)
+        local r = get_request()
+        name_ffi.data = name
+        name_ffi.len = #name
+        local ret =
```
Comment (chronolaw) [inline lualib/resty/kong/request.lua:25]: `ret` is too normal, use a better one.

File: src/ngx_http_lua_kong_request.c [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+    for (size_t i = 0; i < name.len; i++) {
+        hash = ngx_hash(hash, name.data[i]);
+    }
+
+    ngx_http_core_main_conf_t *cmcf =
+        ngx_http_get_module_main_conf(r, ngx_http_core_module);
+    ngx_http_header_t *hh =
+        ngx_hash_find(&cmcf->headers_in_hash, hash, name.data, name.len);
+
+    /* There header is unknown or is not hashed yet. */
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:21]: There header is unknown or not hashed yet

File: src/ngx_http_lua_kong_request.c [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    /* There header is hashed but not cached yet for some reason. */
+    if (hh->offset == 0) {
+        return NULL;
+    }
+
+    /* The header value was already cached in some field of the r->headers_in
+        struct (hh->offset tells in which one). */
+    return &((*(ngx_table_elt_t **)
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:33]: This C style cast is complex, could we format the code easier to read?

File: lualib/resty/kong/request.lua [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+local C = ffi.C
+local ffi_new = ffi.new
+local ffi_str = ffi.string
+
+local subsystem = ngx.config.subsystem
+
+local get_header
+if subsystem == "http" then
+    ffi.cdef[[
+        ngx_str_t *ngx_http_lua_kong_request_get_header(
```
Comment (chronolaw) [inline lualib/resty/kong/request.lua:14]: The function called by lua.ffi should have `ffi` in name, such as `ngx_http_lua_kong_ffi_request_get_header `.

File: src/ngx_http_lua_kong_request.c (hunk 16 above)
Comment (chronolaw) [reply src/ngx_http_lua_kong_request.c:8]: The function's name should have a prefix `ngx_http_`.

File: src/ngx_http_lua_kong_request.c [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+        ((char *) &r->headers_in + hh->offset))->value);
+}
+
+/* linear search */
+static ngx_str_t *search_unknown_header(
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:38]: The function's name should have a prefix` ngx_http_`.

File: src/ngx_http_lua_kong_request.c [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    ngx_http_core_main_conf_t *cmcf =
+        ngx_http_get_module_main_conf(r, ngx_http_core_module);
+    ngx_http_header_t *hh =
+        ngx_hash_find(&cmcf->headers_in_hash, hash, name.data, name.len);
+
+    /* There header is unknown or is not hashed yet. */
+    if (hh == NULL) {
+        return NULL;
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:23]: We can add some `ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ...)` , then check the log in the test case.

File: src/ngx_http_lua_kong_request.c [hunk 30]
```diff
@@ -0,0 +1,109 @@
...
+            i = 0;
+        }
+
+        if (header[i].hash == 0) {
+            continue;
+        }
+
+        size_t n;
+        for (n = 0u; n < name.len && n < header[i].key.len; n++) {
+            u_char ch = ngx_tolower(header[i].key.data[n]);
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:62]: There is a field `lowcase_key` in `ngx_table_elt_t`, maybe we can use it.

File: src/ngx_http_lua_kong_request.c [hunk 31]
```diff
@@ -0,0 +1,114 @@
+#include <ngx_config.h>
+#include <ngx_core.h>
+#include <ngx_http.h>
+
+/* name should be lowercased and preprocessed for both search functions */
+
+/* by hash */
+static ngx_str_t *search_known_header(ngx_http_request_t *r, ngx_str_t name)
+{
+    ngx_http_core_main_conf_t *cmcf;
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:10]: Nginx style, vars should align,

```
xxx     *a;
xx         b;
xxxx     c;
```

File: src/ngx_http_lua_kong_request.c [hunk 32]
```diff
@@ -0,0 +1,114 @@
...
+    ngx_http_header_t *hh;
+    size_t i;
+    ngx_uint_t hash = 0;
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:13]: Add a new line.

File: src/ngx_http_lua_kong_request.c [hunk 33]
```diff
@@ -0,0 +1,114 @@
...
+    /* Calculate a hash of lowercased header name */
+    for (i = 0; i < name.len; i++) {
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:15]: use `ngx_hash_key`?

File: src/ngx_http_lua_kong_request.c (hunk 17 above)
Comment (chronolaw) [reply src/ngx_http_lua_kong_request.c:45]: vars need to align with blank.

File: src/ngx_http_lua_kong_request.c [hunk 34]
```diff
@@ -0,0 +1,114 @@
+#include <ngx_config.h>
```
Comment (chronolaw) [inline src/ngx_http_lua_kong_request.c:1]: Add a file header comment, please refer to other source file.

File: src/ngx_http_lua_kong_request.c (hunk 16 above)
Comment (Suika-Kong) [reply src/ngx_http_lua_kong_request.c:8]: > 

Do we have to add a prefix to local functions? (they are static and in file scope)

File: src/ngx_http_lua_kong_request.c (hunk 21 above)
Comment (Suika-Kong) [reply src/ngx_http_lua_kong_request.c]: > ngx_strlow

I'm also processing underscores so this is not the case.

File: src/ngx_http_lua_kong_request.c (hunk 20 above)
Comment (Suika-Kong) [reply src/ngx_http_lua_kong_request.c:87]: `result` or even `preprocessed_header` would be the same. The meaning of `ret` is clear here as this is what I want to use to store the result of this function.

File: lualib/resty/kong/request.lua (hunk 24 above)
Comment (Suika-Kong) [reply lualib/resty/kong/request.lua:25]: Ditto

File: src/ngx_http_lua_kong_request.c (hunk 25 above)
Comment (Suika-Kong) [reply src/ngx_http_lua_kong_request.c:22]: It literally means the header is unknown or not hashed yet at this place.
But sure, it's more clear to say "the header".

File: src/ngx_http_lua_kong_request.c (hunk 26 above)
Comment (Suika-Kong) [reply src/ngx_http_lua_kong_request.c:34]: Sure. But part of it is still meaningless, so the best I can do is:
```C
    header_found = *(ngx_table_elt_t **)((char *) &r->headers_in + hh->offset);
    return &(header_found->value);
```

Comment (Suika-Kong) [reply src/ngx_http_lua_kong_request.c:34]: Or there would be a headache to name: what does `((char *) &r->headers_in + hh->offset)` ever mean?

File: src/ngx_http_lua_kong_request.c (hunk 28 above)
Comment (Suika-Kong) [reply src/ngx_http_lua_kong_request.c:39]: Ditto

File: src/ngx_http_lua_kong_request.c (hunk 29 above)
Comment (Suika-Kong) [reply src/ngx_http_lua_kong_request.c:24]: Sure. But better from another place.

File: src/ngx_http_lua_kong_request.c (hunk 30 above)
Comment (Suika-Kong) [reply src/ngx_http_lua_kong_request.c]: Notice that we have to deal with `'-'`. But it's still useful.

File: src/ngx_http_lua_kong_request.c (hunk 20 above)
Comment (chronolaw) [reply src/ngx_http_lua_kong_request.c:107]: `ret` is too trivial, just like `i` `n`, it should have more meaning.

File: src/ngx_http_lua_kong_request.c (hunk 16 above)
Comment (chronolaw) [reply src/ngx_http_lua_kong_request.c]: Yes, we can refer nginx source.

File: lualib/resty/kong/request.lua [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+if subsystem == "http" then
+    ffi.cdef[[
+        ngx_str_t *ngx_http_lua_kong_ffi_request_get_header(
+            ngx_http_request_t *r, ngx_str_t name, size_t search_limit);
+    ]]
+
+    local DEFAULT_HEADER_LIMIT = 100
+
+    local name_str = ffi_new("ngx_str_t")
+    get_header = function (name, limit)
```
Comment (fffonion) [inline lualib/resty/kong/request.lua:21]: we should do sanity check on `name` and `limit`.

File: lualib/resty/kong/request.lua [hunk 36]
```diff
@@ -0,0 +1,35 @@
...
+        local r = get_request()
+        name_str.data = name
+        name_str.len = #name
+        local ret = C.ngx_http_lua_kong_ffi_request_get_header(r,
```
Comment (fffonion) [inline lualib/resty/kong/request.lua:25]: we would also need consider the case where an header is specified mutliple times.

File: src/ngx_http_lua_kong_request.c [hunk 37]
```diff
@@ -0,0 +1,166 @@
...
+    size_t               n;
+    u_char               ch;
+    ngx_list_part_t     *part;
+    ngx_table_elt_t     *header;
+
+    part = &(r->headers_in.headers.part);
+    header = part->elts;
+
+    /* not limited when search_limit == 0 */
+    for (i = 0u; search_limit == 0u || i < search_limit; i++) {
```
Comment (fffonion) [inline src/ngx_http_lua_kong_request.c:74]: usually we don't use `0u` but just `0`. is there a reason for this usage?

File: src/ngx_http_lua_kong_request.c [hunk 38]
```diff
@@ -0,0 +1,166 @@
...
+            if (ch == '-') {
+                ch = '_';
+            }
+
+            if (name.data[n] != ch) {
+                break;
+            }
+        }
+
+        if (n == name.len && n == header[i].key.len) {
```
Comment (fffonion) [inline src/ngx_http_lua_kong_request.c:100]: we can do this first before str compare, if length not equal then `continue`

File: src/ngx_http_lua_kong_request.c (hunk 37 above)
Comment (chronolaw) [reply src/ngx_http_lua_kong_request.c]: fix it.

File: lualib/resty/kong/request.lua (hunk 35 above)
Comment (chronolaw) [reply lualib/resty/kong/request.lua:23]: done

File: lualib/resty/kong/request.lua (hunk 36 above)
Comment (chronolaw) [reply lualib/resty/kong/request.lua]: Could you explain it in detail? What shall we do?

**Comment by chronolaw** [comment]
We did a simple performance test, found this PR can not beat `ngx.req.get_headers` and `ngx.var.http_xxx`.
So we decide to pause this PR.

Test data:
1.6640000343323
0.83300018310547
2.4869999885559

curl:
curl 127.1:8080 -H "aaa: 1" -H "bbb: 2" -H "ccc: 3" -H "ddd: 4" -H "content-type: text"

Test code:

```
-- in content_by_lua_block
local last_time = 0

local function timeit()
    ngx.update_time()
    local a = last_time
    last_time = ngx.now()
    return last_time - a
end

timeit()
local get_headers = ngx.req.get_headers

local ITER = 5000*1000
for i=1,ITER do
    local x = get_headers()["content_type"]
end
ngx.say(timeit())

local ITER = 5000*1000
for i=1,ITER do
    local x = ngx.var.http_content_type
end
ngx.say(timeit())

local get_header = require("resty.kong.request").get_header
local ITER = 5000*1000
for i=1,ITER do
    local x = get_header("content_type")
end
ngx.say(timeit())
```

**Comment by chronolaw** [comment]
I also found a bug in this PR.

if the request has no hashed header, such as `content_type`, then `header_found` in `ngx_http_lua_kong_search_known_header` will be `NULL`, it will cause nginx core dump.

But now we have no reason to fix it, maybe someone else can consider it in the future.

**Pull request closed without merging**